service.unmanic/resources/main.py

```diff
@@ -0,0 +1,123 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+
+"""
+    unmanic
+
+    Written by:               Josh.5 <jsunnex@gmail.com>
+    Date:                     06 Dec 2020, (16:17 AM)
+
+    Copyright:
+        Copyright (C) 2021 Josh Sunnex
+
+        This program is free software: you can redistribute it and/or modify it under the terms of the GNU General
+        Public License as published by the Free Software Foundation, version 3.
+
+        This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the
+        implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
+        for more details.
+
+        You should have received a copy of the GNU General Public License along with this program.
+        If not, see <https://www.gnu.org/licenses/>.
+
+"""
+
+import os
+import sys
+import pprint
+import shutil
+import subprocess
+import xbmc
+import xbmcvfs
+import xbmcaddon
+from resources import kodi_log_pipe
+
+__addon__ = xbmcaddon.Addon()
+__path__ = __addon__.getAddonInfo('path')
+__addonname__ = __addon__.getAddonInfo('name')
+__icon__ = __addon__.getAddonInfo('icon')
+__language__ = __addon__.getLocalizedString
+__profile__ = xbmcvfs.translatePath(__addon__.getAddonInfo('profile'))
+__unmanic_module_path__ = xbmcaddon.Addon('script.module.unmanic').getAddonInfo('path')
+
+# Modify path to include lib directory
+sys.path.append(xbmcvfs.translatePath(os.path.join(__path__, 'resources', 'lib')))
+
+
+class UnmanicServiceHandle(object):
```

Comment:
Are you running a service within a service? Why not just import the lib and run it?

Reply:
Because Kodi shuts down subprocesses too quickly when you shut it. It doesn't give Unmanic enough time to stop the processes and so will hang.
By calling it as a subprocess we can kill the process before Kodi tries to and that way we wont end up with Kodi crapping itself.
If anyone has any suggestions on a better way to do this, I'm all ears.